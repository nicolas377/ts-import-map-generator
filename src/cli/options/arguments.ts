// This is where the code that tells the parser what arguments exist, and how to parse and validate them goes.
// This is the only place where the parser should need to be modified to add new arguments.

export type ArgumentType = string | boolean | number;

type StringFromArgumentType<T extends ArgumentType> = T extends string
  ? "string"
  : T extends boolean
  ? "boolean"
  : T extends number
  ? "number"
  : never;

interface BaseArgumentData<T extends ArgumentType> {
  id: ArgumentKind;
  name: string;
  description: string;
  required: boolean;
  type: StringFromArgumentType<T>;
  defaultValue: T;
  singleDashNames: string[];
  doubleDashNames: string[];
  dataGetter: (value: string) => T | Error;
  validator: (value: T) => boolean;
}

interface StringArgumentType extends BaseArgumentData<string> {
  type: "string";
  defaultValue: string;
  singleDashNames: [];
  dataGetter: (value: string) => string | Error;
  validator: (value: string) => boolean;
}

interface BooleanArgumentType extends BaseArgumentData<boolean> {
  type: "boolean";
  defaultValue: boolean;
  singleDashNames: string[];
  dataGetter: (value: string) => boolean | Error;
  validator: (value: boolean) => boolean;
}

interface NumberArgumentType extends BaseArgumentData<number> {
  type: "number";
  defaultValue: number;
  singleDashNames: [];
  dataGetter: (value: string) => number | Error;
  validator: (value: number) => boolean;
}

type ArgumentData =
  | StringArgumentType
  | BooleanArgumentType
  | NumberArgumentType;

export const idToDataMap = new Map<ArgumentKind, ArgumentData>();
export const nameToIdMap = new Map<
  string,
  [singleOrDoubleDash: "single" | "double", kind: ArgumentKind]
>();
export const requiredArgumentIds: ArgumentKind[] = [];

function createArgument(data: ArgumentData): void {
  const { id, singleDashNames, doubleDashNames, required } = data;

  idToDataMap.set(id, data);
  if (required) {
    requiredArgumentIds.push(id);
  }
  for (const name of singleDashNames) {
    nameToIdMap.set(name, ["single", id]);
  }
  for (const name of doubleDashNames) {
    nameToIdMap.set(name, ["double", id]);
  }
}

export const missingArgumentSymbol = Symbol.for("missing");

function numberDataGetter(value: string): number {
  return Number.parseInt(value, 10);
}

function booleanDataGetter(value: string): boolean | Error {
  if (value === "true") {
    return true;
  }
  if (value === "false") {
    return false;
  }
  return new Error(`Invalid boolean value: ${value}`);
}

function noopValidator() {
  return true;
}

// This is where the code that tells the parser what arguments exist, and how to parse and validate them goes.

export const enum ArgumentKind {
  Help,
  Version,
  Entrypoint,
  GraphMaxDepth,
  IgnoreFiles,
}

createArgument({
  id: ArgumentKind.Help,
  name: "help",
  description: "Show this help message.",
  required: false,
  type: "boolean",
  defaultValue: false,
  singleDashNames: ["h"],
  doubleDashNames: ["help"],
  dataGetter: booleanDataGetter,
  validator: noopValidator,
});

createArgument({
  id: ArgumentKind.Version,
  name: "version",
  description: "Show the version of the import map generator.",
  required: false,
  type: "boolean",
  defaultValue: false,
  singleDashNames: ["v"],
  doubleDashNames: ["version"],
  dataGetter: booleanDataGetter,
  validator: noopValidator,
});

createArgument({
  id: ArgumentKind.Entrypoint,
  name: "entrypoint",
  description: "The entrypoint to generate an import map from.",
  required: true,
  type: "string",
  defaultValue: missingArgumentSymbol as unknown as string,
  singleDashNames: [],
  doubleDashNames: ["entrypoint"],
  dataGetter: (value) => value,
  validator: (value) => value !== "",
});

createArgument({
  id: ArgumentKind.GraphMaxDepth,
  name: "graph-max-depth",
  description: "The maximum depth of the graph to generate.",
  required: false,
  type: "number",
  defaultValue: 1000,
  singleDashNames: [],
  doubleDashNames: ["graph-max-depth", "max-depth"],
  dataGetter: numberDataGetter,
  validator: (value) => value >= 0,
});

createArgument({
  id: ArgumentKind.IgnoreFiles,
  name: "ignore-files",
  description: "A comma-separated list of globs to ignore when matched.",
  required: false,
  type: "string",
  defaultValue: "",
  singleDashNames: [],
  doubleDashNames: ["ignore-files", "ignore"],
  dataGetter: (value) => value,
  // TODO: Validate that the value is a comma-separated list of valid globs when implementing globs.
  validator: noopValidator,
});
