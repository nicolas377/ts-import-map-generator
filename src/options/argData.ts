// This is where the code that tells the parser what arguments exist, and how to parse and validate them goes.

type ArgumentType = "string" | "boolean" | "number";

type TypeFromArgumentType<T extends ArgumentType> = T extends "string"
  ? string
  : T extends "boolean"
  ? boolean
  : number;

interface ArgumentData<T extends ArgumentType> {
  id: ArgumentKind;
  name: string;
  description: string;
  type: T;
  defaultValue: TypeFromArgumentType<T>;
  singleDashNames: T extends "boolean" ? string[] : [];
  doubleDashNames: string[];
  validator?: (value: TypeFromArgumentType<T>) => boolean;
}

const nameToIdTable: Record<string, ArgumentKind> = {};

function createArgument<T extends ArgumentType>(
  id: ArgumentKind,
  name: string,
  description: string,
  type: T,
  defaultValue: TypeFromArgumentType<T>,
  singleDashNames: T extends "boolean" ? string[] : [],
  doubleDashNames: string[],
  /** For strings, you can use RegExp.prototype.test */
  validator?: (value: TypeFromArgumentType<T>) => boolean
): ArgumentData<T> {
  nameToIdTable[name] = id;

  return {
    id,
    name,
    description,
    type,
    singleDashNames,
    doubleDashNames,
    defaultValue,
    validator,
  };
}

const enum ArgumentKind {}
