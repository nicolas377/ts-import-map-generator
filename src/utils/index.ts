export * from "utils/debug";
export * from "utils/generalHelpers";
export * from "utils/options";
