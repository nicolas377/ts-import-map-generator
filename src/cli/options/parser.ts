import {
  scanArgs,
  ScannerNodeKind,
  TextNodeWithFlags as ScannerTextNode,
  WhitespaceNode as ScannerWhitespaceNode,
} from "./scanner";
import { Debug } from "utils";

const enum NodeKind {
  SyntaxTree,
  Separator,
  Argument,
  Flag,
  Value,
  EqualsSign,
  Dash,
  UnknownText,
}

export const enum NodeFlags {
  None = 0,
  HasMoreThanTwoDashes = 1 << 0, // This dash node was created from a scanned node of more than two dashes.
  NarrowedFromUnflaggedText = 1 << 1, // This value or flag node was narrowed from an UnknownTextNode by narrowUnknownTextNode().
  IsTopLevelSyntaxTree = 1 << 2, // This node is the full syntax tree.
  ForceCreated = 1 << 3, // This argument node was forcefully created, because we ran into the end of text while in the middle of creating a argument.
}

type NodeId = number & { unique: symbol };

type ExcludeParent<T extends Node> = Omit<T, "parent">;

interface ParentLessNode {
  readonly kind: NodeKind;
  readonly id: NodeId;
  flags: NodeFlags;
}

interface Node extends ParentLessNode {
  readonly parent: ParentLessNode;
}

interface UnknownTextNode extends Node {
  readonly kind: NodeKind.UnknownText;
  readonly text: string;
}

interface SeparatorNode extends Node {
  readonly kind: NodeKind.Separator;
}

interface ValueNode extends Node {
  readonly kind: NodeKind.Value;
  readonly text: string;
}

interface FlagNode extends Node {
  readonly kind: NodeKind.Flag;
  readonly text: string;
}

interface DashNode extends Node {
  readonly kind: NodeKind.Dash;

  readonly singleDash: boolean;
  readonly doubleDash: boolean;
}

export interface ArgumentNode extends Node {
  readonly kind: NodeKind.Argument;

  readonly dash: DashNode;
  readonly flag: FlagNode;
  separator?: SeparatorNode;
  value?: ValueNode;
}

interface SyntaxTree extends ParentLessNode {
  readonly kind: NodeKind.SyntaxTree;
  readonly arguments: ArgumentNode[];
}

interface SyntaxTreeApi {
  readonly syntaxTree: SyntaxTree;
  findNodeById(id: NodeId): ParentLessNode | undefined;
  forEachNode(cb: (node: Node) => void): void;
}

export function parseSyntaxTreeFromArgsString(
  argsString: string
): SyntaxTreeApi {
  // The parser is made to construct a syntax tree.
  // The syntax tree starts with a root node, and then has a series of argument nodes.
  // Each argument node has a dash node, a flag node, an equals sign node, and a value node.
  // The parser is built around parsing those nodes from the phrases yielded by the scanner, and creating the syntax tree from them.
  // There are a few special cases:
  // - If the current dash node has one dash, and the scanner yields an unflagged text node, the the text node should be treated with each individual character as a flag.
  //   This is because short flags can be combined, e.g. -abc is equivalent to -a -b -c.
  // - If the current dash node has one dash, a equals sign node and a value node should be ignored.
  //   This is because short flags cannot have values attached to them.

  const idNodeMap = new Map<NodeId, ParentLessNode>();
  const syntaxTree: SyntaxTree = Object.assign(
    createParentLessNode(NodeKind.SyntaxTree),
    { arguments: [] }
  );
  applyFlagsToNode(syntaxTree, NodeFlags.IsTopLevelSyntaxTree);
  const scannedArgs = [...scanArgs(argsString)];

  let currentScannedArgIndex = -1;
  let currentScannedArg = scannedArgs[currentScannedArgIndex];
  let isOutOfScannedArgs = false as boolean;

  const flagDefaults = {
    singleDash: false,
    doubleDash: false,
    moreThanTwoDashes: false,
  };

  let singleDashFlag: boolean = flagDefaults.singleDash;
  let doubleDashFlag: boolean = flagDefaults.doubleDash;
  let moreThanTwoDashesFlag: boolean = flagDefaults.moreThanTwoDashes;

  let currentDashNode: ExcludeParent<DashNode> | undefined;
  let currentFlagNode: ExcludeParent<FlagNode> | undefined;
  let currentSeparatorNode: ExcludeParent<SeparatorNode> | undefined;
  let currentValueNode: ExcludeParent<ValueNode> | undefined;

  while (currentScannedArgIndex < scannedArgs.length) {
    if (currentScannedArg === undefined) {
      // This is the first iteration. We should just set up the first scanned arg and continue.
      setupNextScannedArg();
      continue;
    }

    // If we encounter whitespace or an equals sign, check if we're creating a long flag argument and the flag is already set.
    if (scannerNodeIsSeparator(currentScannedArg)) {
      // We won't do anything else with a separator node, so we can just discard it if we don't do anything now.
      if (currentFlagNode && isCurrentlyCreatingArgument() && doubleDashFlag) {
        // If we are, we should create and bind a separator node.
        currentSeparatorNode = createParentLessSeparatorNode();
      } else if (!currentFlagNode) {
        Debug.log.warn("Unexpected whitespace or equals sign in argument.");
      }
    }

    // If we encounter dashes, check if we are already creating an argument.
    else if (currentScannedArg.dashFlag) {
      if (!isCurrentlyCreatingArgument()) {
        // If we aren't, create and bind a dash node.
        singleDashFlag = currentScannedArg.text.length === 1;
        doubleDashFlag = currentScannedArg.text.length >= 2;
        moreThanTwoDashesFlag = currentScannedArg.text.length > 2;

        currentDashNode = createParentLessDashNode();
      } else {
        Debug.log.warn("Unexpected dashes in argument.");
      }
    }

    // If the current dash node has one dash, then the text node should be treated with each individual character as a flag.
    else if (singleDashFlag) {
      for (const char of currentScannedArg.text) {
        currentDashNode = createParentLessDashNode();
        currentFlagNode = createParentLessFlagNode(char);
        createAndBindArgumentNode();
      }

      resetForNewArgument();
      setupNextScannedArg();
      continue;
    }

    // If we encounter an unflagged text node, create a UnknownTextNode.
    else {
      const unknownTextNode = createParentLessUnknownTextNode(
        currentScannedArg.text
      );

      // Narrow it to a FlagNode or a ValueNode (offloaded to another function), then bind it.
      const narrowedTextNodeOrFalse = narrowUnknownTextNode(unknownTextNode);

      if (narrowedTextNodeOrFalse === false) {
        // Do nothing. We don't want to bind the node.
        Debug.log.warn("Unknown text node.");
      } else if (narrowedTextNodeOrFalse.kind === NodeKind.Flag) {
        // If we can narrow the unknown text node to a flag node, we should create and bind a flag node.
        currentFlagNode = narrowedTextNodeOrFalse;
      } else {
        // If we can narrow the unknown text node to a value node, we should create and bind a value node.
        currentValueNode = narrowedTextNodeOrFalse;
      }
    }

    // Check if everything that's needed for a argument node has been created.
    // If it has, then create the argument node, bind it to the syntax tree, and clean up for the next node.
    // If there's no more left scanned nodes left, and we can't naturally make an argument node, then check if we can forcefully create one.
    // We need a dash node and a flag node to forcefully create a argument node.
    if (
      canNaturallyCreateArgumentNode() ||
      (isOutOfScannedArgs && currentDashNode && currentFlagNode)
    ) {
      createAndBindArgumentNode();
      resetForNewArgument();
    }

    if (isOutOfScannedArgs) {
      break;
    }

    // Otherwise, just keep going.
    setupNextScannedArg();
  }

  return finalizeSyntaxTreeAndCreateApi(syntaxTree, idNodeMap);

  // Attempts to narrow the unknown text node to a flag node or a value node depending on the context.
  function narrowUnknownTextNode(
    unknownTextNode: ExcludeParent<UnknownTextNode>
  ): ExcludeParent<FlagNode> | ExcludeParent<ValueNode> | false {
    // If we're not currently creating an argument, then we can't narrow the unknown text node, so we log a warning and return false.
    if (!isCurrentlyCreatingArgument()) {
      Debug.log.warn("Unexpected text node between arguments.");
      return false;
    }

    const oneNodeForward = peekForwardNArgs(1);
    const twoNodesForward = peekForwardNArgs(2);

    // Narrowing based on context.

    // If the next token is a equals sign, then we should narrow the unknown text node to a flag node.
    if (scannerNodeIsText(oneNodeForward) && oneNodeForward.equalsFlag) {
      return createFlagNode();
    }
    // If two tokens forward is a dash node, or the previous node is a separator node, then we should narrow the unknown text node to a value node.
    const oneNodeBack = peekForwardNArgs(-1);
    if (
      (scannerNodeIsText(twoNodesForward) &&
        twoNodesForward.dashFlag &&
        // We can't narrow to a value node if the previous node is the start of a new argument (dashes).
        !(scannerNodeIsText(oneNodeBack) && oneNodeBack.dashFlag)) ||
      scannerNodeIsSeparator(oneNodeBack)
    ) {
      return createValueNode();
    }
    // If we need to fill a hole at currentFlagNode, then we should create a flag node.
    if (currentFlagNode === undefined) {
      return createFlagNode();
    }
    // If we need to fill a hole at currentValueNode, then we should create a value node.
    if (currentValueNode === undefined) {
      return createValueNode();
    }

    return false;

    function createFlagNode(): ExcludeParent<FlagNode> {
      const node = createParentLessFlagNode(unknownTextNode.text);
      applyFlagsToNode(node, NodeFlags.NarrowedFromUnflaggedText);

      return node;
    }

    function createValueNode(): ExcludeParent<ValueNode> {
      const node = createParentLessValueNode(unknownTextNode.text);
      applyFlagsToNode(node, NodeFlags.NarrowedFromUnflaggedText);

      return node;
    }
  }

  function canNaturallyCreateArgumentNode(): boolean {
    let canCreateArgument = false;

    // We need a dash node and a flag node no matter what.
    if (currentDashNode && currentFlagNode) {
      // If the dash node has two dashes, we can have a separator and value node, so check ahead to see if those are present.
      if (doubleDashFlag) {
        const forward2Args = peekForwardNArgs(2);

        // If two arguments forward is the start of a new argument or is the end of the string, this argument ends now.
        // Note that we don't support "--flag= --nextFlag", so we don't need to check for that.
        if (scannerNodeIsText(forward2Args) && forward2Args.dashFlag) {
          canCreateArgument = true;
        }
      }

      // If the dash node has one dash, we can't have a separator or value node, so we can create an argument node.
      else if (singleDashFlag) {
        canCreateArgument = true;
      }
    }

    return canCreateArgument;
  }

  function isCurrentlyCreatingArgument(): boolean {
    return currentDashNode !== undefined;
  }

  function createAndBindArgumentNode(): NodeId {
    const argumentNode = createArgumentNode(
      currentDashNode!,
      currentFlagNode!,
      currentSeparatorNode,
      currentValueNode
    );
    syntaxTree.arguments.push(argumentNode);

    return argumentNode.id;
  }

  function createArgumentNode(
    dash: ExcludeParent<DashNode>,
    flag: ExcludeParent<FlagNode>,
    separator?: ExcludeParent<SeparatorNode>,
    value?: ExcludeParent<ValueNode>
  ): ArgumentNode {
    const argumentNode: ArgumentNode = Object.assign(
      createParentLessNode(NodeKind.Argument),
      {
        parent: syntaxTree,
        dash: dash as DashNode,
        flag: flag as FlagNode,
      }
    );

    Object.assign(dash, { parent: argumentNode });
    Object.assign(flag, { parent: argumentNode });

    if (value && separator) {
      argumentNode.separator = Object.assign(separator, {
        parent: argumentNode,
      });
      argumentNode.value = Object.assign(value, { parent: argumentNode });
    }

    return argumentNode;
  }

  function createParentLessDashNode(): ExcludeParent<DashNode> {
    return Object.assign(createParentLessNode(NodeKind.Dash), {
      flags: moreThanTwoDashesFlag
        ? NodeFlags.HasMoreThanTwoDashes
        : NodeFlags.None,
      singleDash: singleDashFlag,
      doubleDash: doubleDashFlag,
    });
  }

  function createParentLessSeparatorNode(): ExcludeParent<SeparatorNode> {
    return createParentLessNode(NodeKind.Separator);
  }

  function createParentLessUnknownTextNode(
    text: string
  ): ExcludeParent<UnknownTextNode> {
    return createTextNode(NodeKind.UnknownText, text);
  }

  function createParentLessValueNode(text: string): ExcludeParent<ValueNode> {
    return createTextNode(NodeKind.Value, text);
  }

  function createParentLessFlagNode(text: string): ExcludeParent<FlagNode> {
    return createTextNode(NodeKind.Flag, text);
  }

  function createTextNode<T extends NodeKind>(kind: T, text: string) {
    return Object.assign(createParentLessNode(kind), { text });
  }

  function createParentLessNode<T extends NodeKind>(kind: T) {
    const node = {
      kind,
      flags: NodeFlags.None,
      id: createNodeId(),
    };

    idNodeMap.set(node.id, node);

    return node;
  }

  function createNodeId(): NodeId {
    let i = 0 as NodeId;

    while (idNodeMap.has(i)) {
      i++;
    }

    return i;
  }

  // helper functions here

  function setupNextScannedArg(): void {
    const nextScannedArgIndex = currentScannedArgIndex + 1;

    // If the next scanned arg index is out of bounds, then we are out of scanned args after this one.
    if (nextScannedArgIndex >= scannedArgs.length - 1) {
      isOutOfScannedArgs = true;
    }

    currentScannedArgIndex = nextScannedArgIndex;
    currentScannedArg = scannedArgs[currentScannedArgIndex];
  }

  function resetForNewArgument(): void {
    currentDashNode = undefined;
    currentFlagNode = undefined;
    currentSeparatorNode = undefined;
    currentValueNode = undefined;

    singleDashFlag = flagDefaults.singleDash;
    doubleDashFlag = flagDefaults.doubleDash;
    moreThanTwoDashesFlag = flagDefaults.moreThanTwoDashes;
  }

  // narrowing functions here

  function nodeIsArgument(node: Node): node is ArgumentNode {
    return node.kind === NodeKind.Argument;
  }

  function scannerNodeIsText(
    node: ScannerTextNode | ScannerWhitespaceNode | undefined
  ): node is ScannerTextNode {
    return node?.kind === ScannerNodeKind.Text;
  }

  function scannerNodeIsSeparator(
    node: ScannerTextNode | ScannerWhitespaceNode | undefined
  ): node is ScannerWhitespaceNode | (ScannerTextNode & { equalsFlag: true }) {
    return (
      (node?.kind === ScannerNodeKind.Whitespace || node?.equalsFlag) ?? false
    );
  }

  function peekForwardNArgs(
    n: number
  ): ScannerTextNode | ScannerWhitespaceNode | undefined {
    return scannedArgs[currentScannedArgIndex + n];
  }

  function applyFlagsToNode(node: ParentLessNode, flags: NodeFlags): void {
    node.flags |= flags;
  }

  function finalizeSyntaxTreeAndCreateApi(
    syntaxTree: SyntaxTree,
    idNodeMap: Map<NodeId, ParentLessNode>
  ): SyntaxTreeApi {
    pruneIdNodeMap();

    return {
      syntaxTree,
      findNodeById(id: NodeId): ParentLessNode | undefined {
        return idNodeMap.get(id);
      },
      forEachNode: forEachNodeInSyntaxTree,
    };

    function pruneIdNodeMap(): void {
      const idsInSyntaxTree: NodeId[] = [];

      forEachNodeInSyntaxTree((node) => {
        idsInSyntaxTree.push(node.id);
      });

      for (const id of idNodeMap.keys()) {
        if (!idsInSyntaxTree.includes(id)) {
          idNodeMap.delete(id);
        }
      }
    }

    function forEachNodeInSyntaxTree(callback: (node: Node) => void): void {
      // We start at the top of the syntax tree and traverse down to the bottom, calling the callback on each node.
      traverseDown(syntaxTree);

      function traverseDown(node: Node | SyntaxTree): void {
        if (node.kind === NodeKind.SyntaxTree) {
          (node as SyntaxTree).arguments.forEach(traverseDown);
          return;
        }

        callback(node);

        if (nodeIsArgument(node)) {
          traverseDown(node.dash);
          traverseDown(node.flag);
          if (node.separator) {
            traverseDown(node.separator);
          }
          if (node.value) {
            traverseDown(node.value);
          }
        }
      }
    }
  }
}
