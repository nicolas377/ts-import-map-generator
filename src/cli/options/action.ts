import {
  ArgumentKind,
  ArgumentType,
  idToDataMap,
  nameToIdMap,
} from "./arguments";
import { parseSyntaxTreeFromArgsString } from "./parser";
import { Debug, programOptions } from "utils";

interface ActionToTake {
  id: ArgumentKind;
  value: ArgumentType;
}

export function takeActionFromCliArgs(argsString: string): void {
  // This is the real worker of the argument parser.
  // It takes the string of arguments, parses it, and then takes the appropriate action based on the parsed arguments.
  // We want everything to be as modular as possible, so we rely on data provided in arguments.ts to do the argument-specific work.
  // The goal of this function is to set everything in programOptions before starting the actual work.

  const syntaxTreeApi = parseSyntaxTreeFromArgsString(argsString);

  const actionsToTake: ActionToTake[] = [];

  for (const treeArgument of syntaxTreeApi.syntaxTree.arguments) {
    const argumentFlagName = treeArgument.flag.text;
    // Try to parse the argument's id from its name. If it can't be parsed, then it's an unknown argument and should be ignored.
    const [singleOrDoubleDash, argumentId] =
      nameToIdMap.get(argumentFlagName) ?? [];
    const optionDetails = idToDataMap.get(argumentId!);
    const { dash: argumentDash, value: argumentValue } = treeArgument;

    if (
      singleOrDoubleDash === undefined ||
      argumentId === undefined ||
      optionDetails === undefined ||
      // The argument is known, but has the wrong dash.
      (singleOrDoubleDash === "single" && argumentDash.singleDash === false) ||
      (singleOrDoubleDash === "double" && argumentDash.doubleDash === false)
    ) {
      // The argument is unknown, so we should ignore it.
      Debug.log.warn(`Unknown argument: ${argumentFlagName}`);
      continue;
    }

    // If the argument is a boolean argument, then it doesn't have a value.
    // If there's no value, then set the value to the default value.
    // If there is a value, then set the value to the parsed value.
    if (optionDetails.type === "boolean") {
      const value =
        argumentValue === undefined
          ? true
          : optionDetails.dataGetter(argumentValue.text);

      if (value instanceof Error) {
        Debug.log.warn(`Invalid value for argument: ${argumentFlagName}`);
        continue;
      }

      if (optionDetails.validator(value)) {
        if (value !== optionDetails.defaultValue)
          addAction(optionDetails.id, value);
      } else {
        Debug.log.warn(`Invalid value for argument: ${argumentFlagName}`);
      }
    } else {
      if (argumentValue === undefined) {
        Debug.log.warn(`Missing value for argument: ${argumentFlagName}`);
        continue;
      }

      // The argument is known and has a value, so we should parse it and take the appropriate action.
      const value = optionDetails.dataGetter(argumentValue.text);

      if (value instanceof Error) {
        Debug.log.warn(`Invalid value for argument: ${argumentFlagName}`);
        continue;
      }

      if (
        (optionDetails.validator as (value: string | number) => boolean)(
          value
        ) &&
        value !== optionDetails.defaultValue
      ) {
        addAction(optionDetails.id, value);
      } else {
        Debug.log.warn(
          `Invalid value ${argumentValue.text} for argument: ${argumentFlagName}`
        );
      }
    }
  }

  for (const { id, value } of actionsToTake) {
    programOptions.setOption(id, value);
  }

  function addAction(id: ArgumentKind, value: ArgumentType): void {
    actionsToTake.push({
      id,
      value,
    });
  }
}
