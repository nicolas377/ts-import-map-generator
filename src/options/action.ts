import { idToDataMap, nameToIdMap } from "./arguments";
import { NodeFlags, parseSyntaxTreeFromArgsString } from "./parser";

type OptionValue = string | boolean | number;

interface ActionToTake {
  name: string;
  value: OptionValue;
  isDefault: boolean;
}

export function takeActionFromCliArgs(argsString: string): void {
  // This is the real worker of the argument parser.
  // It takes the string of arguments, parses it, and then takes the appropriate action based on the parsed arguments.
  // We want everything to be as modular as possible, so we rely on data provided in arguments.ts to do the argument-specific work.
  // The goal of this function is to set everything in programOptions before starting the actual work.

  const syntaxTreeApi = parseSyntaxTreeFromArgsString(argsString);

  const actionsToTake: ActionToTake[] = [];

  for (const treeArgument of syntaxTreeApi.syntaxTree.arguments) {
    // Try to parse the argument's id from its name. If it can't be parsed, then it's an unknown argument and should be ignored.
    const argumentId = nameToIdMap.get(treeArgument.flag.text);

    if (argumentId === undefined) {
      continue;
    }

    // The argument is known, so we can get its data from the map.
    const optionDetails = idToDataMap.get(argumentId)!;
    const { value: argumentValue } = treeArgument;

    // TODO: pick up here
  }

  console.log(actionsToTake);
}
