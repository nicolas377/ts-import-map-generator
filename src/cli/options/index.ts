import { argv } from "node:process";
import { takeActionFromCliArgs } from "./action";

export function initializeOptionsFromCliArgs(): void {
  const argsString: string = argv.slice(2).join(" ");

  takeActionFromCliArgs(argsString);
}
