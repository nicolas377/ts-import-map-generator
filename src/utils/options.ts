import { ArgumentKind, ArgumentType, idToDataMap } from "options/arguments";

const argumentKindToOptionNameMap: Record<ArgumentKind, keyof IOptions> = {
  [ArgumentKind.Help]: "printHelpAndExit",
  [ArgumentKind.Version]: "printVersionAndExit",
};

interface IOptions {
  setOption(kind: ArgumentKind, value: ArgumentType): void;
  printHelpAndExit: boolean;
  printVersionAndExit: boolean;
}

class OptionsClass {
  private allOptions = [ArgumentKind.Help, ArgumentKind.Version] as const;

  private options: Record<ArgumentKind, ArgumentType>;

  public setOption(kind: ArgumentKind, value: ArgumentType): void {
    (this[
      argumentKindToOptionNameMap[kind] as keyof OptionsClass
    ] as unknown as ArgumentType) = value;
  }

  constructor() {
    // kind should be one of all of the options, so this validates that at compile-time.
    this.options = this.allOptions.reduce((acc, kind: ArgumentKind) => {
      const { defaultValue } = idToDataMap.get(kind) ?? {};

      if (defaultValue === undefined) {
        throw new Error(`No default value for argument kind: ${kind}`);
      }

      acc[kind] = defaultValue;
      return acc;
    }, {} as Record<ArgumentKind, ArgumentType>);

    for (const kind of this.allOptions) {
      const optionData = idToDataMap.get(kind);

      if (optionData === undefined) {
        throw new Error(`No option data for argument kind: ${kind}`);
      }

      Object.defineProperty(this, argumentKindToOptionNameMap[kind], {
        configurable: false,
        enumerable: true,
        get: () => this.options[kind],
        set: (value: ArgumentType) => {
          // TODO: when TS supports merging function types, this assertion can be removed.
          (optionData.validator as (value: ArgumentType) => boolean)(value);
          this.options[kind] = value;
        },
      });
    }
  }
}

export const programOptions = new OptionsClass() as unknown as IOptions;
