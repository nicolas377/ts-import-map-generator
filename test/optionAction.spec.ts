import { removeDateFromLog } from "./helpers";
import { takeActionFromCliArgs } from "options/action";
import { Debug, programOptions } from "utils";

describe("takeActionFromCliArgs", () => {
  afterEach(() => {
    Debug.logs = [];
  });

  it("should take action based on the arguments", () => {
    const argsString = "--help --version";
    takeActionFromCliArgs(argsString);

    expect(programOptions.printHelpAndExit).toBe(true);
    expect(programOptions.printVersionAndExit).toBe(true);

    expect(Debug.logs.map(removeDateFromLog)).toMatchSnapshot();
  });

  const testStrings = [
    "--help",
    "--version",
    "--help --version",
    "--version --help",
    "-hv",
    "--version=false --help=true",
    "--version=false --help=true --version --help",
  ];

  test.each(testStrings)(
    "should take the same action every time '%s'",
    (argsString) => {
      takeActionFromCliArgs(argsString);

      expect(Debug.logs.map(removeDateFromLog)).toMatchSnapshot("logs");
      expect(programOptions).toMatchSnapshot("programOptions");
    }
  );
});
