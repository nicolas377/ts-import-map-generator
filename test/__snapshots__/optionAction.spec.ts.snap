// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`takeActionFromCliArgs should take action based on the arguments 1`] = `[]`;

exports[`takeActionFromCliArgs should take the same action every time '--help --version': logs 1`] = `[]`;

exports[`takeActionFromCliArgs should take the same action every time '--help --version': programOptions 1`] = `
OptionsClass {
  "allOptions": [
    0,
    1,
  ],
  "options": {
    "0": true,
    "1": true,
  },
  "printHelpAndExit": true,
  "printVersionAndExit": true,
}
`;

exports[`takeActionFromCliArgs should take the same action every time '--help': logs 1`] = `[]`;

exports[`takeActionFromCliArgs should take the same action every time '--help': programOptions 1`] = `
OptionsClass {
  "allOptions": [
    0,
    1,
  ],
  "options": {
    "0": true,
    "1": true,
  },
  "printHelpAndExit": true,
  "printVersionAndExit": true,
}
`;

exports[`takeActionFromCliArgs should take the same action every time '--version --help': logs 1`] = `[]`;

exports[`takeActionFromCliArgs should take the same action every time '--version --help': programOptions 1`] = `
OptionsClass {
  "allOptions": [
    0,
    1,
  ],
  "options": {
    "0": true,
    "1": true,
  },
  "printHelpAndExit": true,
  "printVersionAndExit": true,
}
`;

exports[`takeActionFromCliArgs should take the same action every time '--version': logs 1`] = `[]`;

exports[`takeActionFromCliArgs should take the same action every time '--version': programOptions 1`] = `
OptionsClass {
  "allOptions": [
    0,
    1,
  ],
  "options": {
    "0": true,
    "1": true,
  },
  "printHelpAndExit": true,
  "printVersionAndExit": true,
}
`;

exports[`takeActionFromCliArgs should take the same action every time '--version=false --help=true --version --help': logs 1`] = `
[
  {
    "level": 2,
    "message": "Ignoring default value for argument:  version",
  },
]
`;

exports[`takeActionFromCliArgs should take the same action every time '--version=false --help=true --version --help': programOptions 1`] = `
OptionsClass {
  "allOptions": [
    0,
    1,
  ],
  "options": {
    "0": true,
    "1": true,
  },
  "printHelpAndExit": true,
  "printVersionAndExit": true,
}
`;

exports[`takeActionFromCliArgs should take the same action every time '--version=false --help=true': logs 1`] = `
[
  {
    "level": 2,
    "message": "Ignoring default value for argument:  version",
  },
]
`;

exports[`takeActionFromCliArgs should take the same action every time '--version=false --help=true': programOptions 1`] = `
OptionsClass {
  "allOptions": [
    0,
    1,
  ],
  "options": {
    "0": true,
    "1": true,
  },
  "printHelpAndExit": true,
  "printVersionAndExit": true,
}
`;

exports[`takeActionFromCliArgs should take the same action every time '-hv': logs 1`] = `[]`;

exports[`takeActionFromCliArgs should take the same action every time '-hv': programOptions 1`] = `
OptionsClass {
  "allOptions": [
    0,
    1,
  ],
  "options": {
    "0": true,
    "1": true,
  },
  "printHelpAndExit": true,
  "printVersionAndExit": true,
}
`;
